src/deps-index.ts

```diff
@@ -21,79 +21,100 @@ export interface DepsIndex {
-      const lazyImports = new Set<UrlString>();
-      for (const htmlImport of imports) {
-        try {
-          console.assert(
-              htmlImport.url, 'htmlImport: %s has no url', htmlImport);
-        } catch (err) {
-          if (err instanceof AssertionError) {
-            continue;
-          }
-          throw err;
-        }
+/**
+ * For a given URL, return a set of eager and lazy dependencies.  The
+ * function is recursive, but a `visited` set is carried forward to short-
+ * circuit loops encountered.
+ */
+function getTransitiveDependencies(
+    url: UrlString,
+    analysis: Analysis,
+    visited: Set<UrlString> = new Set<UrlString>()): DependencyMapEntry {
+  const map: DependencyMapEntry = {
+    eager: new Set(),
+    lazy: new Set(),
+    lazyEdges: new Set()
+  };
+  const document = analysis.getDocument(url);
+  if (!(document instanceof Document)) {
+    const message = document && document.message || 'unknown';
+    throw new Error(`Unable to get document ${url}: ${message}`);
+  }
 
-        switch (htmlImport.type) {
-          case 'html-import':
-            eagerImports.add(htmlImport.document.url);
-            break;
-          case 'lazy-html-import':
-            lazyImports.add(htmlImport.document.url);
-            break;
-        }
-      }
-      return {url: url, eager: eagerImports, lazy: lazyImports};
+  // Gets a collection of only the direct html imports for the document.
+  const htmlImports = document.getFeatures(
+      {kind: 'html-import', imported: false, externalPackages: true});
+
+  // Initially, we will place each import into the dependency map based on
+  // whether it was explicitly a lazy import.
+  htmlImports.forEach((i) => {
+    if (i.lazy) {
+      map.lazy.add(i.document.url);
+      map.lazyEdges.add(i.document.url);
+    } else {
+      map.eager.add(i.document.url);
     }
+  });
+
+  visited.add(url);
+
+  // Every unvisited direct eager dependency's lazy dependencies are recognized
+  // as transitively lazy unless already recognized as eager.  All others are
+  // treated as eager.
+  Array.from(map.eager).filter((d) => !visited.has(d)).forEach((d) => {
+    const deps = getTransitiveDependencies(d, analysis, visited);
+    deps.eager.forEach((dep) => map.eager.add(dep) && map.lazy.delete(dep));
+    deps.lazy.forEach((dep) => map.eager.has(dep) || map.lazy.add(dep));
+    deps.lazyEdges.forEach((dep) => map.lazyEdges.add(dep));
+  });
+
+  // Every unvisited direct lazy dependency's dependencies are recognized
+  // as transitively lazy unless already defined as eager.
+  Array.from(map.lazy).filter((d) => !visited.has(d)).forEach((d) => {
+    const deps = getTransitiveDependencies(d, analysis, visited);
+    deps.eager.forEach(
+        (dep) => map.eager.has(dep) ? map.lazy.delete(dep) : map.lazy.add(dep));
+    deps.lazy.forEach((dep) => map.eager.has(dep) || map.lazy.add(dep));
+    deps.lazyEdges.forEach((dep) => map.lazyEdges.add(dep));
+  });
```

Comment:
This algorithm isn't obviously correct to me. Here's my thinking based on what I've inferred from what we're trying to do here, lemmy know if I go off track:

A file has eager and lazy imports of other files. A file is an eager dependency of another if that file is reachable by any path of entirely eager imports. A file is a lazy dependency of another if all paths to that file include at least one lazy import.

This means that every eager dependency of an eager dependency is an eager dependency, and the rest of its dependencies are lazy. So I think an easier to understand implementation would involve building the set of files reachable by by entirely eager paths, and another set of all files reachable at all, then constructing the lazy files as those that are in the set of all files but not in the set of eager files.

I think this algorithm is equivalent in complexity but simpler to reason about.